Improve `GeoJSONSource#updateData` support for string feature ids and reduce likelihood of race conditions

lucaswoj opened the pull request:
Fix https://github.com/maplibre/maplibre-gl-js/issues/6694
Ref https://github.com/maplibre/maplibre-gl-js/issues/4364
Ref #6562 

This PR contains two small fixes for `GeoJSONSource#updateData`

 - It casts `string` feature ids to `number`, if possible. Per #1043 support for non-numeric string feature IDs in GeoJSON is limited. However support for_numeric_ string feature IDs was broken for `GeoJSONSource#updateData` primer to this PR. This PR also adds a warning when using `GeoJSONSource#updateData` with non-numeric string feature IDs. 
 - It reduces the likelihood of race conditions when calling `GeoJSONSource#updateData` while the source is in the middle of another update. 

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] ~Include before/after visuals or gifs if this PR includes visual changes.~
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented on src/source/geojson_source.ts:
```diff
@@ -448,7 +448,14 @@ export class GeoJSONSource extends Evented implements Source {
+        );
+
+        if (prevIds.has(NaN)) {
+            warnOnce('GeoJSONSource#updateData only supports numeric feature IDs or string IDs that can be cast to numbers. See issue #1043 for more details.');
```
A link to the issue would be better I believe from a DX perspective.

wayofthefuture commented:
So we are completely breaking GeoJSON source so that shouldReloadTile works?

HarelM commented:
@wayofthefuture can you maybe add a test case that you would like to make sure is working after all these fixes/changes?
In theory this test should pass prior to the changes related to geojson optimization, right?

wayofthefuture commented:
With the shouldReloadTile function, GeoJSON sources will not update all tiles. Polygons will be cut off. So in order for a small performance gain this happens to those of us using strings:


https://github.com/user-attachments/assets/53dddcc6-3690-42af-95d6-ebc046c748a2

HarelM commented:
I'm guessing this can fallback to previous behavior when IDs are strings, can't it?

lucaswoj commented:
> I'm guessing this can fallback to previous behavior when IDs are strings, can't it?

Sure. That's possible and a reasonable measure for backwards compatibility. 

Do you still want a warning message in that case? I am wary of officially supporting string IDs here while so many other parts of MapLibre don't support them.

HarelM commented:
I think the warning should stay, yes.

lucaswoj commented:
> Would it be possible to use something like https://www.npmjs.com/package/xxhashjs to convert strings to numbers?
> 
> 
> 
> ```ts
> 
> function stringToId(str: string): number {
> 
>   return XXH.h32(str, 0xABCD).toNumber();
> 
> }
> 
> ```
> 
> 
> 
> Or something like that.

Unfortunately it's not so easy. String->i32 hashing produces frequent collisions so we'd need to implement a hash collision strategy too. Not impossible, but way beyond the scope of this bugfix.

lucaswoj commented:
This is ready for another round of review.

 - I have added logic to detect the use of string feature IDs and fall back to the unoptimized behavior. 
 - I looked around and discovered that MapLibre nor `vt-pbf` tries to cast string ids to number ids, so I removed the use of `parseInt` here. Strings are strings and are not supported.

HarelM commented:
I think the warning can be improved a bit to help developers find the problem faster (if they accidentally somehow didn't intent for string IDs).
But otherwise this looks good. 
The above comment is not gating for this PR.

lucaswoj commented on src/source/geojson_source.ts:
```diff
@@ -450,10 +450,19 @@ export class GeoJSONSource extends Evented implements Source {
 
+        for (const id of prevIds.values()) {
+            if (typeof id !== 'number') {
+                warnOnce('GeoJSONSource#updateData is slower when using string GeoJSON feature IDs. Consider using numeric IDs for better performance.');
```
Good idea. Done in add286768

0xFA11 commented:
@wayofthefuture 
> @0xFA11 possible solutions I am considering:
> 
> 1. Wait for MLT.
> 2. Reduce the id string length to about 4-5 characters and hash to numbers on the fly. (Must fit in 32bit)

I think we're happy to switch from MVT to MLT at some point in the near future, we like it already!

for id string hashing, I saw this comment from @lucaswoj 

> Unfortunately it's not so easy. String->i32 hashing produces frequent collisions so we'd need to implement a hash collision strategy too. Not impossible, but way beyond the scope of this bugfix.

we are actually using https://www.npmjs.com/package/js-xxhash package instead of https://www.npmjs.com/package/xxhashjs and `js-xxhash` is performing a bit better in our case (slightly better performance, better typescript, newer than xxhashjs: 8 years vs 1 year)

having said that, xxhash is a pretty solid non-cryptographic hashing algorithm with a good uniform hash distribution and fast performance (even without SIMD). our workaround for not being able to use string IDs in GeoJson features array was literally to do `data.features.push({ id: xxHash32("some-id-here"), type: "Feature", geometry: { ... }, properties: { ... }, ... });` and we had no issues so far after that. only downside there was to also xxHash32 an ID when we wanted to query or update features based on their IDs, not huge deal.

FWIW, I'd say `js-xxhash` even with strings longer than 4-5 characters wouldn't be too bad and people wouldn't bump into hash collisions left and right. we had ~1 million GeoJson points with unique string IDs (like GUIDs/UUIDs) and xxHash'ing them was perfectly fine, we had no collisions whatsoever.

lucaswoj commented:
The max feature id allowed by the Mapbox tile spec is only 2^52. So the likelihood of a collision on 1 million features with a 52 bit hash is small but not insignificant:

```
1 - e^(-(1,000,000)^2/(2×2^53)) =0.00555%
```

Careful! If your dataset continues to grow, the probability continues to rise. At 100 million features, the probability is about 43%. 

And careful! This assumes a perfectly random distribution hash values. In reality the values aren't quite randomly distributed, so this number is only a lower bound on the probability. The actual probability could be much higher.

It's up to you if these risks are acceptable for your application.

lucaswoj commented:
> I guess my ultimate question would be what is the maximum number of lowercase alpha characters that you can uniquely convert to a number that fits in 32bit with no collisions

4

0xFA11 commented:
> The max feature id allowed by the Mapbox tile spec is only 2^52. So the likelihood of a collision on 1 million features with a 52 bit hash is small but not insignificant:
> 
> ```
> 1 - e^(-(1,000,000)^2/(2×2^53)) =0.00555%
> ```
> 
> Careful! If your dataset continues to grow, the probability continues to rise. At 100 million features, the probability is about 43%.
> 
> And careful! This assumes a perfectly random distribution hash values. In reality the values aren't quite randomly distributed, so this number is only a lower bound on the probability. The actual probability could be much higher.
> 
> It's up to you if these risks are acceptable for your application.

we have a top-level 1.2 million cap, it won't grow beyond that but I will do more testing and propose switching from xxhash32 to xxhash64 hash numbers in our specific case. we don't pack data in MVT tiles, we actually set data on map layers/features at runtime progressively while streaming in.

nevertheless, this has been a very good reminder and educational.
